### PR TITLE
1453 create application

### PR DESCRIPTION
**Pull request by BCerki**
Card: https://github.com/bcgov/cas-cif/issues/1453

### CODE REVIEW

**Review by tmastrom — COMMENTED** [review]
Good slice! I'm wondering if we could get the jsonSchemaForm from the database?

**Comment by BCerki** [comment]
> Good slice! I'm wondering if we could get the jsonSchemaForm from the database?

I thought about it, but the fields will probably change a lot throughout the slices, and if the schema's in the db, every time we want to update it we'll need to do a sqitch migration. And doing 10 sqitch migrations will probably be more time-consuming than refactoring it later

File: app/components/Form/ProjectRfpForm.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import { useRouter } from "next/router";
+import { useState } from "react";
+import { graphql, useFragment } from "react-relay";
+import {
+  getExternalProjectRevisionFormPageRoute,
+  getProjectRevisionFormPageRoute,
+} from "routes/pageRoutes";
+
+interface Props {
+  query: any;
```
Comment (dleard) [inline app/components/Form/ProjectRfpForm.tsx:17]: Let's replace this `any` type. Maybe something like this?
```suggestion
  query: ProjectRfpForm_query$key;
```

File: app/lib/theme/widgets/SelectParentWidget.tsx [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { useEffect, useMemo, useState } from "react";
 import { WidgetProps } from "@rjsf/core";
 
 export interface EntitySchema {
-  list: [{ rowId: any }];
+  list: [{ rowId: any; year?: number }];
```
Comment (dleard) [inline app/lib/theme/widgets/SelectParentWidget.tsx:6]: Remove `any` type
```suggestion
  list: [{ rowId: number; year?: number }];
```

File: app/lib/theme/widgets/SelectParentWidget.tsx [hunk 3]
```diff
@@ -13,34 +13,38 @@ interface SelectParentComponentProps extends WidgetProps {
   parent: EntitySchema;
   child: EntitySchema;
   foreignKey: string;
+  displayIfChildrenEmpty: any;
```
Comment (dleard) [inline app/lib/theme/widgets/SelectParentWidget.tsx:16]: Remove `any` type
```suggestion
  displayIfChildrenEmpty: string;
```

File: app/pages/cif/project-revision/[projectRevision]/form/[formIndex]/index.tsx [hunk 4]
```diff
@@ -3,17 +3,9 @@ import { Button } from "@button-inc/bcgov-theme";
...
-import useRedirectToLatestRevision from "hooks/useRedirectToLatestRevision";
-import useRedirectToValidFormIndex from "hooks/useRedirectToValidFormIndex";
 import withRelayOptions from "lib/relay/withRelayOptions";
-import { useCreateProjectRevision } from "mutations/ProjectRevision/createProjectRevision";
 import { useRouter } from "next/router";
-import {
-  getProjectRevisionFormPageRoute,
-  getProjectRevisionPageRoute,
-} from "routes/pageRoutes";
+import { useFormIndexHelpers } from "../../../../../../hooks/useFormIndexHelpers";
```
Comment (dleard) [inline app/pages/cif/project-revision/[projectRevision]/form/[formIndex]/index.tsx:8]: We shouldn't need all the ../../.. here
```suggestion
import { useFormIndexHelpers } from "hooks/useFormIndexHelpers";
```

**Review by dleard — CHANGES_REQUESTED** [review]

**Comment by tmastrom** [comment]
> 

That's a really good plan while things are in flux

File: app/cypress/support/commands.js [hunk 5]
```diff
@@ -84,9 +84,13 @@ Cypress.Commands.add(
   "fillAndCheckNewProjectForm",
   (fundingStream, fundingStreamYear) => {
     cy.url().should("include", "/new");
-    cy.findByLabelText(/Funding Stream$/i).select(fundingStream);
-    cy.findByLabelText(/RFP Year/i).select(fundingStreamYear);
-
+    cy.get("[id=select-parent-dropdown-root_fundingStreamRfpId]")
+      // selecting twice ensures cypress selects the correct funding stream
+      .select(fundingStream)
+      .select(fundingStream);
```
Comment (tmastrom) [inline app/cypress/support/commands.js:90]: I'm not able to notice a difference in testing with just one or two of these statements. Interestingly, asserting that the funding stream should be visible after it is selected causes an error `cy.contains(fundingStream).should("be.visible"); // error`

File: app/pages/cif-external/project-revision/[projectRevision]/form/[formIndex]/index.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    }
+  }
+`;
+
+export function ExternalProjectFormPage({
+  preloadedQuery,
+}: RelayProps<{}, FormIndexExternalPageQuery>) {
+  const { query } = usePreloadedQuery(pageQuery, preloadedQuery);
+
+  console.log(
```
Comment (dleard) [inline app/pages/cif-external/project-revision/[projectRevision]/form/[formIndex]/index.tsx:50]: Remove console log

File: app/pages/cif-external/project-revision/[projectRevision]/form/[formIndex]/index.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    "query.projectRevision?.projectId",
+    query.projectRevision?.projectId
+  );
+
+  let mode: TaskListMode;
+  if (!query.projectRevision?.projectId) mode = "create";
+  else if (query.projectRevision.changeStatus === "committed") mode = "view";
+  else mode = "update";
+  console.log("mode", mode);
```
Comment (dleard) [inline app/pages/cif-external/project-revision/[projectRevision]/form/[formIndex]/index.tsx:59]: Remove console log

File: app/routes/pageRoutes.ts [hunk 8]
```diff
@@ -61,21 +61,30 @@ export const getNewProjectRevisionPageRoute = () => ({
...
+};
 
 export const getProjectRevisionFormPageRoute = (
   projectRevisionId: string,
   formIndex: string | number,
   anchor: string = undefined,
+  isInternal: boolean = true,
   isRoutedFromNew?: boolean
 ) => {
+  console.log("is internal", isInternal);
```
Comment (dleard) [inline app/routes/pageRoutes.ts:84]: Remove console log

**Review by dleard — CHANGES_REQUESTED** [review]
A couple things I found during manual testing:

Clicking submit application overview after filling out the Project Overview & Legal Name directs me to 404 page not found.

If "Return to Dashboard" is clicked while the project is in draft, there is no way to get back to that draft

**Comment by BCerki** [comment]
> A couple things I found during manual testing:
> 
> Clicking submit application overview after filling out the Project Overview & Legal Name directs me to 404 page not found.
> 
> If "Return to Dashboard" is clicked while the project is in draft, there is no way to get back to that draft

The next pages (attachments, summary) aren't part of this ticket, so that's why you get a 404 page. Getting back to the draft/edit also isn't part of this ticket (or the slice).

**Pull request merged**